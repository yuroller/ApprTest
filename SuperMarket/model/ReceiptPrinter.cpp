#include "ReceiptPrinter.h"

namespace {
    std::string getFormattedNumberAsString(double number, int precision)
    {
        std::stringstream stream;
        stream << std::fixed << std::setprecision(precision) << number;
        return stream.str();
    }
}

ReceiptPrinter::ReceiptPrinter() : ReceiptPrinter(40)
{
}

ReceiptPrinter::ReceiptPrinter(int columns) : columns(columns)
{}

std::string ReceiptPrinter::printReceipt(const Receipt& receipt)
{
    std::string result;
    for (const auto& item : receipt.getItems())
    {
        result.append(presentReceiptItem(item));
    }
    for (const auto& discount : receipt.getDiscounts())
    {
        result.append(presentDiscount(discount));
    }
    result.append("\n");
    result.append(presentTotal(receipt));
    return result;
}

std::string ReceiptPrinter::presentReceiptItem(const ReceiptItem& item) const
{
    std::string price = getFormattedNumberAsString(item.getTotalPrice(), 2);
    std::string name = item.getProduct().getName();

    std::string line = formatLineWithWhitespace(name, price);

    if (item.getQuantity() != 1)
    {
        line += "  " + getFormattedNumberAsString(item.getPrice(), 2) + " * " + presentQuantity(item) + "\n";
    }
    return line;
}

std::string ReceiptPrinter::presentDiscount(const Discount& discount) const
{
    std::string name = discount.getDescription() + "(" + discount.getProduct().getName() + ")";
    std::string pricePresentation = getFormattedNumberAsString(discount.getDiscountAmount(), 2);
    return formatLineWithWhitespace(name, pricePresentation);
}

std::string ReceiptPrinter::presentTotal(const Receipt& receipt) const
{
    std::string total = "Total: ";
    std::string pricePresentation = presentPrice(receipt.getTotalPrice());
    return formatLineWithWhitespace(total, pricePresentation);
}

std::string ReceiptPrinter::formatLineWithWhitespace(const std::string& name, const std::string& value) const
{
    int whitespaceSize = columns - name.length() - value.length();
    std::string whitespace;
    for (int i = 0; i < whitespaceSize; i++)
    {
        whitespace.append(" ");
    }
    return name + whitespace + value + "\n";
}

std::string ReceiptPrinter::presentPrice(double price) const
{
    return getFormattedNumberAsString(price, 2);
}


std::string ReceiptPrinter::presentQuantity(const ReceiptItem& item)
{
    return ProductUnit::Each == item.getProduct().getUnit()
        ? getFormattedNumberAsString(item.getQuantity(), 0)
        : getFormattedNumberAsString(item.getQuantity(), 3);
}