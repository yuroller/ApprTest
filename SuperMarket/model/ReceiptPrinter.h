#ifndef CPP_RECEIPTPRINTER_H
#define CPP_RECEIPTPRINTER_H


#include "Receipt.h"

#include <iomanip>
#include <sstream>

class ReceiptPrinter
{
public:
    ReceiptPrinter();
    ReceiptPrinter(int columns);
    std::string printReceipt(const Receipt& receipt);
    std::string presentReceiptItem(const ReceiptItem& item) const;
    std::string presentDiscount(const Discount& discount) const;
    std::string presentTotal(const Receipt& receipt) const;
    std::string formatLineWithWhitespace(const std::string& name, const std::string& value) const;
    std::string presentPrice(double price) const;
    static std::string presentQuantity(const ReceiptItem& item);

private:
    const int columns;
};


#endif //CPP_RECEIPTPRINTER_H
