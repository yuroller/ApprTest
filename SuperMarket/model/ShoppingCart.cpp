#include "ShoppingCart.h"

std::vector<ProductQuantity> ShoppingCart::getItems() const {
    return items;
}

std::map<Product, double> ShoppingCart::getProductQuantities() const {
    return productQuantities;
}

void ShoppingCart::addItem(const Product& product) {
    addItemQuantity(product, 1.0);
}

void ShoppingCart::addItemQuantity(const Product& product, double quantity) {
    items.emplace_back(product, quantity);
    if (productQuantities.find(product) != productQuantities.end()) {
        productQuantities[product] += quantity;
    } else {
        productQuantities[product] = quantity;
    }
}

void ShoppingCart::handleOffers(Receipt& receipt, std::map<Product, std::unique_ptr<Offer>> &offers, SupermarketCatalog* catalog) {
    for (const auto& productQuantity : productQuantities) {
        Product product = productQuantity.first;
        if (offers.find(product) != offers.end()) {
            Discount* discount = offers[product]->handleOffer(catalog, productQuantities);
            if (discount != nullptr)
                receipt.addDiscount(*discount);
        }
    }
}