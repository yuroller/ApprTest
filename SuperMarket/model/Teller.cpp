#include "Teller.h"

Teller::Teller(SupermarketCatalog *catalog) : catalog(catalog) {}

void Teller::addSpecialOffer(SpecialOfferType offerType, const Product& product, double argument) {
    auto offer = [&]() -> Offer* {
        switch (offerType) {
        case SpecialOfferType::ThreeForTwo:
            return new BuyXQtyPayYQtyOffer(product, 3, 2);
        case SpecialOfferType::TenPercentDiscount:
            return new PercentDiscountOffer(product, argument);
        case SpecialOfferType::TwoForAmount:
            return new BuyXQtyPayAmountOffer(product, 2, argument);
        case SpecialOfferType::FiveForAmount:
            return new BuyXQtyPayAmountOffer(product, 5, argument);
        }

        return nullptr;
    }();
    
    if (offer != nullptr)
        offers.emplace(product, offer);
}

Receipt Teller::checksOutArticlesFrom(ShoppingCart theCart) {
    Receipt receipt{};
    std::vector<ProductQuantity> productQuantities = theCart.getItems();
    for (const auto& pq: productQuantities) {
        Product p = pq.getProduct();
        double quantity = pq.getQuantity();
        double unitPrice = catalog->getUnitPrice(p);
        double price = quantity * unitPrice;
        receipt.addProduct(p, quantity, unitPrice, price);
    }
    theCart.handleOffers(receipt, offers, catalog);

    return receipt;
}
