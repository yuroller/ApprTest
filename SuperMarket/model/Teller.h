#ifndef CPP_TELLER_H
#define CPP_TELLER_H


#include <map>
#include <memory>
#include "SupermarketCatalog.h"
#include "Offer.h"
#include "Receipt.h"
#include "ShoppingCart.h"
#include "SpecialOfferType.h"

class Teller {
public:
    Teller(SupermarketCatalog* catalog);

    void addSpecialOffer(SpecialOfferType offerType, const Product& product, double argument);

    Receipt checksOutArticlesFrom(ShoppingCart theCart);

private:
    SupermarketCatalog* catalog;
    std::map<Product, std::unique_ptr<Offer>> offers;
};


#endif //CPP_TELLER_H
