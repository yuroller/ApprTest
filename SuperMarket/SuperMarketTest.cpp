#define APPROVAL_TESTS_HIDE_DEPRECATED_CODE 1
#include "ApprovalTests.hpp"
#include "catch2/catch.hpp"

#include "model/SupermarketCatalog.h"
#include "FakeCatalog.h"
#include "model/ShoppingCart.h"
#include "model/Teller.h"
#include "model/ReceiptPrinter.h"


using namespace ApprovalTests;

TEST_CASE("SingleProductReceipt", "[Supermarket]")
{
    // ARRANGE
    SupermarketCatalog *catalog = new FakeCatalog();
    Product toothbrush("toothbrush", ProductUnit::Each);
    catalog->addProduct(toothbrush, 0.99);
    Product apples("apples", ProductUnit::Kilo);
    catalog->addProduct(apples, 1.99);
    Product soap("soap", ProductUnit::Each);
    catalog->addProduct(soap, 0.59);
    Product shampoo("shampoo", ProductUnit::Each);
    catalog->addProduct(shampoo, 1.19);

    Teller teller(catalog);
    teller.addSpecialOffer(SpecialOfferType::TenPercentDiscount, toothbrush, 10.0);
    teller.addSpecialOffer(SpecialOfferType::ThreeForTwo, soap, 0.0);
    teller.addSpecialOffer(SpecialOfferType::TwoForAmount, shampoo, 2.0);

    ShoppingCart cart;
    cart.addItem(toothbrush);
    cart.addItemQuantity(apples, 2.5);
    cart.addItem(toothbrush);
    cart.addItemQuantity(soap, 3);
    cart.addItemQuantity(shampoo, 2);

    // ACT
    Receipt receipt = teller.checksOutArticlesFrom(cart);

    // Print Compare/Approve
    ReceiptPrinter printer;
    Approvals::verify(printer.printReceipt(receipt));
}



