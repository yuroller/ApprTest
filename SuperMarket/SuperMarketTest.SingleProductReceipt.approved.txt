toothbrush                          0.99
apples                              4.97
  1.99 * 2.500
toothbrush                          0.99
10.000000% off(toothbrush)         -0.20

Total:                              6.76

